### PR TITLE
Add data, ui lattices and ui primitives for the discuss context

### PR DESCRIPTION
**Pull request by rtshkmr**
## Motivation
This PR defines some functions that are used when defining the two lattices that shall be kept within the socket -- data lattice and ui_lattice. 

Additionally, I've also added some bottom up ui primitives that we will be building on top of for the next bit when displaying discussions.

Note the following:
1. I'm currently looking only at the :index action for discussions
2. there's no pagination for now, pagination shall be left for later -- until the various moving parts get merged together

## Screenshots 
image

image

### CODE REVIEW

File: lib/vyasa/sangh/sheaf_lattice.ex
Comment (rtshkmr) [inline lib/vyasa/sangh/sheaf_lattice.ex:33]: QQ: @ks0m1c is it okay to preload marks when creating the sheaf?

for now, i'm creating a complete sheaf lattice, while ignoring the pagination piece. Will ignore pagination until things are more ready.

File: docs/crudrice.livemd
Comment (rtshkmr) [inline docs/crudrice.livemd:222]: To get the meat of this PR, referring to the livebook from this section onwards should be sufficient.

File: lib/vyasa/sangh/sheaf_lattice.ex
Comment (rtshkmr) [inline lib/vyasa/sangh/sheaf_lattice.ex:73]: Here's the main accessor helper function to read from the lattice. My next PR shall be relying heavily on this. 

The examples above show the various cases I expect this read function to be used for.

File: lib/vyasa_web/components/contexts/ui_state.ex
Comment (rtshkmr) [inline lib/vyasa_web/components/contexts/ui_state.ex:7]: the ui state for sheaf is a container that encapsulates the ui state of a collection of marks

**Review by rtshkmr — COMMENTED** [review]
This PR defines some functions that are used when defining the two lattices that shall be kept within the socket -- data lattice and ui_lattice

Note the following:
1. I'm currently looking only at the :index action for discussions
2. there's no pagination for now, pagination shall be left for later -- until the various moving parts get merged together

File: lib/vyasa/sangh/sheaf_lattice.ex
Comment (ks0m1c) [reply lib/vyasa/sangh/sheaf_lattice.ex:33]: should preload at get_root_sheafs_by_session/1

File: lib/vyasa/sangh/sheaf_lattice.ex
Comment (ks0m1c) [reply lib/vyasa/sangh/sheaf_lattice.ex:73]: HAHAHA this filter pattern is quite cool create_sheaf_lattice_filter could easily use 
for bot get_in and get_and_update_in behav

[Access.filter(&(create_sheaf_lattice_filter(&1)))]

File: lib/vyasa_web/components/contexts/components.ex
Comment (ks0m1c) [inline lib/vyasa_web/components/contexts/components.ex:394]: probably should be sheaf body as title and first mark

File: lib/vyasa/sangh/sheaf_lattice.ex
Comment (rtshkmr) [reply lib/vyasa/sangh/sheaf_lattice.ex:71]: I'll carry on first, will leave it to you to make this change yeah

**Comment by rtshkmr** [comment]
@ks0m1c give an approval if this is ready to merge, i want to merge the ready PRs in 

I don't intent to commit anything new to this

File: lib/vyasa/sangh/sheaf_lattice.ex
Comment (ks0m1c) [reply lib/vyasa/sangh/sheaf_lattice.ex:71]: you want me to commit this in(?)

File: lib/vyasa/sangh/sheaf_lattice.ex
Comment (rtshkmr) [reply lib/vyasa/sangh/sheaf_lattice.ex]: done this, marking this as resolved

File: lib/vyasa_web/components/contexts/components.ex
Comment (rtshkmr) [reply lib/vyasa_web/components/contexts/components.ex:387]: you're right, will get back to this when i'm doing ui iteration again after the current immediate next set of tasks

**Review by ks0m1c — APPROVED** [review]

**Pull request merged**
